<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>API &#8212; XYalign 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Release History" href="release.html" />
    <link rel="prev" title="Frequently Asked Questions" href="faqs.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="api">
<h1>API<a class="headerlink" href="#api" title="Permalink to this headline">¶</a></h1>
<div class="section" id="reference-fasta-files">
<h2>Reference/Fasta files<a class="headerlink" href="#reference-fasta-files" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="reftools.RefFasta">
<em class="property">class </em><code class="descclassname">reftools.</code><code class="descname">RefFasta</code><span class="sig-paren">(</span><em>filepath</em>, <em>samtools='samtools'</em>, <em>bwa='bwa'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta" title="Permalink to this definition">¶</a></dt>
<dd><p>A class for working with external reference fasta files</p>
<p class="rubric">Attributes</p>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="86%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>filepath</td>
<td>(str) Full path to external bam file.</td>
</tr>
<tr class="row-even"><td>samtools</td>
<td>(str) Full path to samtools. Default = &#8216;samtools&#8217;</td>
</tr>
<tr class="row-odd"><td>bwa</td>
<td>(str) Full path to bwa. Default = &#8216;bwa&#8217;</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="reftools.RefFasta.check_bwa_index">
<code class="descname">check_bwa_index</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.check_bwa_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.check_bwa_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks to see if bwa indices are newer than fasta.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">bool</p>
<blockquote class="last">
<div><p>True if indices exist and are newer than fasta. False otherwise.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.chromosome_lengths">
<code class="descname">chromosome_lengths</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.chromosome_lengths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p>Chromosome lengths ordered by sequence order in fasta</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.chromosome_names">
<code class="descname">chromosome_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.chromosome_names"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.chromosome_names" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p>Chromosome names ordered by sequence order in fasta</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.conditional_index_bwa">
<code class="descname">conditional_index_bwa</code><span class="sig-paren">(</span><em>bwa='bwa'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.conditional_index_bwa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.conditional_index_bwa" title="Permalink to this definition">¶</a></dt>
<dd><p>Indexes if indices are the same age or older than the fasta.
Use RefFasta.index_bwa() to force indexing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>bwa</strong> : str</p>
<blockquote class="last">
<div><p>Path to bwa program (default is &#8216;bwa&#8217;)</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.index_bwa">
<code class="descname">index_bwa</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.index_bwa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.index_bwa" title="Permalink to this definition">¶</a></dt>
<dd><p>Index reference using bwa</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">bool</p>
<blockquote>
<div><p>True if successful</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>RuntimeError</strong></p>
<blockquote class="last">
<div><p>If return code from external call is not 0</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.index_fai">
<code class="descname">index_fai</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.index_fai"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.index_fai" title="Permalink to this definition">¶</a></dt>
<dd><p>Create fai index for reference using samtools (&#8216;samtools faidx ref.fa&#8217;)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">bool</p>
<blockquote>
<div><p>True if successful</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>RuntimeError</strong></p>
<blockquote class="last">
<div><p>If return code from external call is not 0</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.is_faidxed">
<code class="descname">is_faidxed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.is_faidxed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.is_faidxed" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks that fai index exists, is not empty and is newer than reference.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">bool</p>
<blockquote class="last">
<div><p>True if fai index exists and is newer than fasta, False otherwise.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.isolate_chroms">
<code class="descname">isolate_chroms</code><span class="sig-paren">(</span><em>new_ref_prefix</em>, <em>chroms</em>, <em>bed_mask</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.isolate_chroms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.isolate_chroms" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a reference fasta file and a list of chromosomes to include
and outputs a new, indexed (and optionally masked) reference fasta.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>new_ref_prefix</strong> : str</p>
<blockquote>
<div><p>The desired path to and prefix of the output files</p>
</div></blockquote>
<p><strong>chroms</strong> : list</p>
<blockquote>
<div><p>Chromosomes to include in the output fasta</p>
</div></blockquote>
<p><strong>bed_mask</strong> : str</p>
<blockquote>
<div><p>Bed file of regions to mask (as N) in the new reference</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">str</p>
<blockquote class="last">
<div><p>Path to new, indexed (optionally masked) fasta</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.mask_reference">
<code class="descname">mask_reference</code><span class="sig-paren">(</span><em>bed_mask</em>, <em>output_fasta=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.mask_reference"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.mask_reference" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a new masked references by hardmasking regions included
in the bed_mask</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>bed_mask</strong> : str</p>
<blockquote>
<div><p>Bed file of regions to mask (as N) in the new reference</p>
</div></blockquote>
<p><strong>output_fasta</strong> : str</p>
<blockquote>
<div><p>The full path to and filename of the output fasta</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">str</p>
<blockquote class="last">
<div><p>Path to new (indexed and masked) fasta</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="reftools.RefFasta.seq_dict">
<code class="descname">seq_dict</code><span class="sig-paren">(</span><em>out_dict=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/reftools.html#RefFasta.seq_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#reftools.RefFasta.seq_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Create sequence dictionary .dict file using samtools</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>out_dict</strong> : str</p>
<blockquote>
<div><p>The desired file name for the sequence dictionary -
defaults to adding &#8216;.dict&#8217; to the fasta name</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">bool</p>
<blockquote>
<div><p>True if exit code of external call is 0.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>RuntimeError</strong></p>
<blockquote class="last">
<div><p>If external call exit code is not 0.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="bam-files">
<h2>Bam files<a class="headerlink" href="#bam-files" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="bam.BamFile">
<em class="property">class </em><code class="descclassname">bam.</code><code class="descname">BamFile</code><span class="sig-paren">(</span><em>filepath</em>, <em>samtools='samtools'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#BamFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.BamFile" title="Permalink to this definition">¶</a></dt>
<dd><p>A class for working with external bam files</p>
<p class="rubric">Attributes</p>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="86%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>filepath</td>
<td>(str) Full path to external bam file.</td>
</tr>
<tr class="row-even"><td>samtools</td>
<td>(str) Full path to samtools. Default = &#8216;samtools&#8217;</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="bam.BamFile.analyze_bam_fetch">
<code class="descname">analyze_bam_fetch</code><span class="sig-paren">(</span><em>chrom</em>, <em>window_size</em>, <em>target_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#BamFile.analyze_bam_fetch"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.BamFile.analyze_bam_fetch" title="Permalink to this definition">¶</a></dt>
<dd><p>Analyze BAM (or CRAM) file for depth and mapping quality across genomic
windows.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>chrom</strong> : str</p>
<blockquote>
<div><p>The name of the chromosome to analyze</p>
</div></blockquote>
<p><strong>window_size</strong></p>
<blockquote>
<div><p>If int, the window size to use for sliding window analyses, if None
intervals from target_file</p>
</div></blockquote>
<p><strong>target_file</strong> : str</p>
<blockquote>
<div><p>Path to bed_file containing regions to analyze instead of
windows of a fixed size. Will only be engaged if window_size is None</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">dict</p>
<blockquote class="last">
<div><p>A dictionary in which the key is &#8216;windows&#8217; and the value is
a pandas data frame with the columns: &#8220;chrom&#8221;, &#8220;start&#8221;, &#8220;stop&#8221;,
&#8220;depth&#8221;, &#8220;mapq&#8221;</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="bam.BamFile.chromosome_lengths">
<code class="descname">chromosome_lengths</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#BamFile.chromosome_lengths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.BamFile.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p>chromosome lengths ordered by sequence order in bam header</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="bam.BamFile.chromosome_names">
<code class="descname">chromosome_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#BamFile.chromosome_names"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.BamFile.chromosome_names" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p>chromosome names ordered by sequence order in bam header</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="bam.BamFile.get_chrom_length">
<code class="descname">get_chrom_length</code><span class="sig-paren">(</span><em>chrom</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#BamFile.get_chrom_length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.BamFile.get_chrom_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract chromosome length from BAM header.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>chrom</strong> : str</p>
<blockquote>
<div><p>The name of the chromosome or scaffold.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>length</strong> : int</p>
<blockquote>
<div><p>The length (integer) of the chromsome/scaffold</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>RuntimeError</strong></p>
<blockquote class="last">
<div><p>If chromosome name not present in bam header</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="bam.BamFile.index_bam">
<code class="descname">index_bam</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#BamFile.index_bam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.BamFile.index_bam" title="Permalink to this definition">¶</a></dt>
<dd><p>Indexes a bam using samtools (&#8216;samtools index file.bam&#8217;).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">bool</p>
<blockquote>
<div><p>True if successful.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>RuntimeError</strong></p>
<blockquote class="last">
<div><p>If return code from external call is not 0.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="bam.BamFile.is_indexed">
<code class="descname">is_indexed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#BamFile.is_indexed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.BamFile.is_indexed" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks that bam index exists, is not empty, and is newer than bam.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">bool</p>
<blockquote class="last">
<div><p>True if bam index exists and is newer than bam, False otherwise.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="bam.BamFile.strip_reads">
<code class="descname">strip_reads</code><span class="sig-paren">(</span><em>repairsh</em>, <em>single</em>, <em>output_directory</em>, <em>output_prefix</em>, <em>regions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#BamFile.strip_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.BamFile.strip_reads" title="Permalink to this definition">¶</a></dt>
<dd><p>Strips reads from a bam or cram file in provided regions and outputs
sorted fastqs containing reads, one set of fastq files per read group.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>repairsh</strong> : str</p>
<blockquote>
<div><p>Path to repair.sh (from BBmap)</p>
</div></blockquote>
<p><strong>single</strong> : bool</p>
<blockquote>
<div><p>If true output single-end fastq, otherwise output paired-end fastqs</p>
</div></blockquote>
<p><strong>output_directory</strong> : str</p>
<blockquote>
<div><p>The directory for ALL output (including temporary files)</p>
</div></blockquote>
<p><strong>output_prefix</strong> : str</p>
<blockquote>
<div><p>The name (without path) to use for prefix to output fastqs</p>
</div></blockquote>
<p><strong>regions</strong> : list</p>
<blockquote>
<div><p>regions from which reads will be stripped</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list</p>
<blockquote class="last">
<div><p>A two-item list containing the path to a text file pairing read group
names with associated output fastqs, and a text file containing a
list of &#64;RG lines associated with each read group</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="bam.sambamba_merge">
<code class="descclassname">bam.</code><code class="descname">sambamba_merge</code><span class="sig-paren">(</span><em>sambamba_path</em>, <em>bam_list</em>, <em>output_prefix</em>, <em>threads</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#sambamba_merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.sambamba_merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges bam files using sambamba.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>sambamba_path</strong> : str</p>
<blockquote>
<div><p>The path to sambamba</p>
</div></blockquote>
<p><strong>bam_list</strong> : list</p>
<blockquote>
<div><p>Bam files to be merged. Merging order will match order of this list.</p>
</div></blockquote>
<p><strong>output_prefix</strong> : str</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">str</p>
<blockquote class="last">
<div><p>path to merged bam</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="bam.switch_sex_chromosomes_sambamba">
<code class="descclassname">bam.</code><code class="descname">switch_sex_chromosomes_sambamba</code><span class="sig-paren">(</span><em>samtools_path</em>, <em>sambamba_path</em>, <em>bam_orig</em>, <em>bam_new</em>, <em>sex_chroms</em>, <em>output_directory</em>, <em>output_prefix</em>, <em>threads</em>, <em>pg_header_dict</em>, <em>cram=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bam.html#switch_sex_chromosomes_sambamba"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bam.switch_sex_chromosomes_sambamba" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes sex chromosomes from original bam and merges in remmapped
sex chromosomes, while retaining the original bam header (and adding new
&#64;PG line)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>samtools_path</strong> : str</p>
<blockquote>
<div><p>The path to samtools</p>
</div></blockquote>
<p><strong>sambamba_path :</strong></p>
<blockquote>
<div><p>The path to sambamba</p>
</div></blockquote>
<p><strong>bam_orig</strong> : str</p>
<blockquote>
<div><p>The path to the original full bam file</p>
</div></blockquote>
<p><strong>bam_new</strong> : str</p>
<blockquote>
<div><p>The path to the bam file containing the remapped sex chromosomes</p>
</div></blockquote>
<p><strong>sex_chroms</strong> : list</p>
<blockquote>
<div><p>Sex chromosomes (to be removed from bam_orig)</p>
</div></blockquote>
<p><strong>output_directory</strong> : str</p>
<blockquote>
<div><p>The path to directory where all files (inc. temp) will be output</p>
</div></blockquote>
<p><strong>output_prefix</strong> : str</p>
<blockquote>
<div><p>The name (without path) to use as prefix for all files</p>
</div></blockquote>
<p><strong>threads</strong> : int</p>
<blockquote>
<div><p>The number of threads/cpus to use</p>
</div></blockquote>
<p><strong>pg_header_dict</strong> : dict</p>
<blockquote>
<div><dl class="docutils">
<dt>dictionary with information to be included in the new &#64;PG line</dt>
<dd><ul class="first last">
<li><dl class="first docutils">
<dt>must contain:</dt>
<dd><p class="first last">Key = &#8216;CL&#8217;, value = list of command line values
Key = &#8216;ID&#8217;, value = string of program ID</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>optional:</dt>
<dd><p class="first last">Key = &#8216;VN&#8217;, value = string of program version</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</div></blockquote>
<p><strong>cram</strong> : bool</p>
<blockquote>
<div><p>If True, will treat input as cram files and output cram files.
Otherwise, will treat input as bam.  Defaule is False. True is currently
unsupported.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">str</p>
<blockquote>
<div><p>Bam or cram file path with new sex chromosomes, but all others intact.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>RuntimeError</strong></p>
<blockquote class="last">
<div><p>If cram is not False.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="variant-calling-vcf-processing">
<h2>Variant Calling/ VCF Processing<a class="headerlink" href="#variant-calling-vcf-processing" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="variants.platypus_caller">
<code class="descclassname">variants.</code><code class="descname">platypus_caller</code><span class="sig-paren">(</span><em>platypus_path</em>, <em>log_path</em>, <em>bam</em>, <em>ref</em>, <em>chroms</em>, <em>cpus</em>, <em>output_file</em>, <em>regions_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants.html#platypus_caller"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#variants.platypus_caller" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses platypus to make variant calls on provided bam file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>platypus_path</strong> : str</p>
<blockquote>
<div><p>Path to platypus</p>
</div></blockquote>
<p><strong>log_path</strong> : str</p>
<blockquote>
<div><p>Path to and name of desired log file for platypus</p>
</div></blockquote>
<p><strong>bam</strong> : str</p>
<blockquote>
<div><p>Path to input bam (or cram) file</p>
</div></blockquote>
<p><strong>ref</strong> : str</p>
<blockquote>
<div><p>Path to reference sequence</p>
</div></blockquote>
<p><strong>chroms</strong> : list</p>
<blockquote>
<div><p>Chromosomes to call variants on, e.g., [&#8220;chrX&#8221;, &#8220;chrY&#8221;, &#8220;chr19&#8221;]</p>
</div></blockquote>
<p><strong>cpus</strong> : int</p>
<blockquote>
<div><p>Number of threads/cores to use</p>
</div></blockquote>
<p><strong>output_file</strong> : path</p>
<blockquote>
<div><p>Path to and name of the output vcf</p>
</div></blockquote>
<p><strong>regions_file</strong> : {str, None}</p>
<blockquote>
<div><p>If not None, must be path to bed file containing regions to call variants
in.  If None, calls in call regions of provided chromosomes. Default =
None.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">int</p>
<blockquote class="last">
<div><p>Exit code of the platypus call</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="variants.parse_platypus_VCF">
<code class="descclassname">variants.</code><code class="descname">parse_platypus_VCF</code><span class="sig-paren">(</span><em>filename</em>, <em>qual_cutoff</em>, <em>chrom</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants.html#parse_platypus_VCF"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#variants.parse_platypus_VCF" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse vcf generated by Platypus to grab read balance</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>filename</strong> : str</p>
<blockquote>
<div><p>Full path to the input vcf</p>
</div></blockquote>
<p><strong>qual_cutoff</strong> : str</p>
<blockquote>
<div><p>Minimum (PHRED) quality at which sites should be included</p>
</div></blockquote>
<p><strong>chrom</strong> : str</p>
<blockquote>
<div><p>Name of the chromosome to include</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><dl class="docutils">
<dt>three corresponding arrays of the same length:</dt>
<dd><p class="first last">(position across the chromosome, site quality, read balance)</p>
</dd>
</dl>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="variants.plot_variants_per_chrom">
<code class="descclassname">variants.</code><code class="descname">plot_variants_per_chrom</code><span class="sig-paren">(</span><em>chrom_list</em>, <em>vcf_file</em>, <em>sampleID</em>, <em>output_prefix</em>, <em>qual_cutoff</em>, <em>MarkerSize</em>, <em>MarkerAlpha</em>, <em>bamfile_obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants.html#plot_variants_per_chrom"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#variants.plot_variants_per_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses a vcf file and plots read balance in separate plots
for each chromosome in the input list</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>chrom_list</strong> : list</p>
<blockquote>
<div><p>Chromosomes to include</p>
</div></blockquote>
<p><strong>vcf_file</strong> : str</p>
<blockquote>
<div><p>File (including path) of platypus vcf to analyze</p>
</div></blockquote>
<p><strong>sampleID</strong> : str</p>
<blockquote>
<div><p>Sample ID (for plot titles)</p>
</div></blockquote>
<p><strong>output_prefix</strong> : str</p>
<blockquote>
<div><p>Full path to and prefix of desired output plots</p>
</div></blockquote>
<p><strong>qual_cutoff</strong> : int</p>
<blockquote>
<div><p>Minimum (Phred) quality to consider a site in the vcf</p>
</div></blockquote>
<p><strong>MarkerSize</strong> : float</p>
<blockquote>
<div><p>Size of markers (matplotlib sizes) to use in the figure</p>
</div></blockquote>
<p><strong>MarkerAlpha</strong> : float</p>
<blockquote>
<div><p>Transparency (matplotlib values, 0 to 1) of markers</p>
</div></blockquote>
<p><strong>bamfile_obj</strong> : BamFile() object</p>
<blockquote>
<div><p>Used to get chromosome lengths only</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">int</p>
<blockquote class="last">
<div><p>0</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="variants.hist_read_balance">
<code class="descclassname">variants.</code><code class="descname">hist_read_balance</code><span class="sig-paren">(</span><em>chrom</em>, <em>readBalance</em>, <em>sampleID</em>, <em>output_prefix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants.html#hist_read_balance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#variants.hist_read_balance" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots a histogram of read balance</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>chrom</strong> : str</p>
<blockquote>
<div><p>Name of the chromosome</p>
</div></blockquote>
<p><strong>readBalance</strong> : nump array</p>
<blockquote>
<div><p>Read balance values</p>
</div></blockquote>
<p><strong>sampleID</strong> : str</p>
<blockquote>
<div><p>Sample name or id to include in the plot title</p>
</div></blockquote>
<p><strong>output_prefix</strong> : str</p>
<blockquote>
<div><p>Desired prefix (including full path) of the output files</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">int</p>
<blockquote class="last">
<div><p>0</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="variants.plot_read_balance">
<code class="descclassname">variants.</code><code class="descname">plot_read_balance</code><span class="sig-paren">(</span><em>chrom</em>, <em>positions</em>, <em>readBalance</em>, <em>sampleID</em>, <em>output_prefix</em>, <em>MarkerSize</em>, <em>MarkerAlpha</em>, <em>bamfile_obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/variants.html#plot_read_balance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#variants.plot_read_balance" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots read balance at each SNP along a chromosome</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>chrom</strong> : str</p>
<blockquote>
<div><p>Name of the chromosome</p>
</div></blockquote>
<p><strong>positions</strong> : numpy array</p>
<blockquote>
<div><p>Positions along the chromosome (same length as readBalance)</p>
</div></blockquote>
<p><strong>readBalance</strong> : numpy array</p>
<blockquote>
<div><p>Read balance corresponding with the positions in the positions array</p>
</div></blockquote>
<p><strong>sampleID</strong> : str</p>
<blockquote>
<div><p>Sample name or id to include in the plot title</p>
</div></blockquote>
<p><strong>output_prefix</strong> : str</p>
<blockquote>
<div><p>Desired prefix (including full path) of the output files</p>
</div></blockquote>
<p><strong>MarkerSize</strong> : float</p>
<blockquote>
<div><p>Size of markers (matplotlib sizes) to use in the figure</p>
</div></blockquote>
<p><strong>MarkerAlpha</strong> : float</p>
<blockquote>
<div><p>Transparency (matplotlib values) of markers for the figure</p>
</div></blockquote>
<p><strong>bamfile_obj</strong> : BamFile() object</p>
<blockquote>
<div><p>Used to get chromosome lengths only</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">int</p>
<blockquote class="last">
<div><p>0</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="ploidy-estimation">
<h2>Ploidy Estimation<a class="headerlink" href="#ploidy-estimation" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="ploidy.bootstrap">
<code class="descclassname">ploidy.</code><code class="descname">bootstrap</code><span class="sig-paren">(</span><em>data_frame</em>, <em>first_chrom</em>, <em>second_chrom</em>, <em>chrom_column</em>, <em>value_column</em>, <em>num_reps</em>, <em>output_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ploidy.html#bootstrap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#ploidy.bootstrap" title="Permalink to this definition">¶</a></dt>
<dd><p>Bootstraps the 95 percent confidence interval of the mean ratio of
measure for two chromosomes (chrom1 / chrom2).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>data_frame</strong> : pandas dataframe</p>
<p><strong>first_chrom</strong> : str</p>
<blockquote>
<div><p>The name of the first chromosome in comparison</p>
</div></blockquote>
<p><strong>second_chrom</strong> : str</p>
<blockquote>
<div><p>The name of the second chromosome in comparison</p>
</div></blockquote>
<p><strong>chrom_column</strong> : str</p>
<blockquote>
<div><p>The name of the column containing chromosome names</p>
</div></blockquote>
<p><strong>value_column</strong> : str</p>
<blockquote>
<div><p>The name of the column containing the value of interest</p>
</div></blockquote>
<p><strong>num_perms</strong> : int</p>
<blockquote>
<div><p>The number of permutations to use</p>
</div></blockquote>
<p><strong>output_file</strong> : {str, None}</p>
<blockquote>
<div><p>If not None, will print results to this file.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p>(0.025 percentile, 0.5 percentile, 0.975 percentile)</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="ploidy.ks_two_sample">
<code class="descclassname">ploidy.</code><code class="descname">ks_two_sample</code><span class="sig-paren">(</span><em>data_frame</em>, <em>first_chrom</em>, <em>second_chrom</em>, <em>chrom_column</em>, <em>value_column</em>, <em>output_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ploidy.html#ks_two_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#ploidy.ks_two_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs a Two-sample Kolmogorov-Smirnov test</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>data_frame</strong> : pandas dataframe</p>
<p><strong>first_chrom</strong> : str</p>
<blockquote>
<div><p>The name of the first chromosome in comparison</p>
</div></blockquote>
<p><strong>second_chrom</strong> : str</p>
<blockquote>
<div><p>The name of the second chromosome in comparison</p>
</div></blockquote>
<p><strong>chrom_column</strong> : str</p>
<blockquote>
<div><p>The name of the column containing chromosome names</p>
</div></blockquote>
<p><strong>value_column</strong> : str</p>
<blockquote>
<div><p>The name of the column containing the value of interest</p>
</div></blockquote>
<p><strong>num_perms</strong> : int</p>
<blockquote>
<div><p>The number of permutations to use</p>
</div></blockquote>
<p><strong>output_file</strong> : {str, None}</p>
<blockquote>
<div><p>If not None, will print results to this file.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p>(ks_statistic, ks_pvalue)</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="ploidy.permutation_test_chromosomes">
<code class="descclassname">ploidy.</code><code class="descname">permutation_test_chromosomes</code><span class="sig-paren">(</span><em>data_frame</em>, <em>first_chrom</em>, <em>second_chrom</em>, <em>chrom_column</em>, <em>value_column</em>, <em>num_perms</em>, <em>output_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ploidy.html#permutation_test_chromosomes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#ploidy.permutation_test_chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs a permutation test comparing mean values of two chromosomes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>data_frame</strong> : pandas dataframe</p>
<p><strong>first_chrom</strong> : str</p>
<blockquote>
<div><p>The name of the first chromosome in comparison</p>
</div></blockquote>
<p><strong>second_chrom</strong> : str</p>
<blockquote>
<div><p>The name of the second chromosome in comparison</p>
</div></blockquote>
<p><strong>chrom_column</strong> : str</p>
<blockquote>
<div><p>The name of the column containing chromosome names</p>
</div></blockquote>
<p><strong>value_column</strong> : str</p>
<blockquote>
<div><p>The name of the column containing the value of interest</p>
</div></blockquote>
<p><strong>num_perms</strong> : int</p>
<blockquote>
<div><p>The number of permutations to use</p>
</div></blockquote>
<p><strong>output_file</strong> : {str, None}</p>
<blockquote>
<div><p>If not None, will print results to this file</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p>(mean of first chrom, mean of second chrom, p-value)</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="mapping-assembly">
<h2>Mapping/assembly<a class="headerlink" href="#mapping-assembly" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="assemble.bwa_mem_mapping_sambamba">
<code class="descclassname">assemble.</code><code class="descname">bwa_mem_mapping_sambamba</code><span class="sig-paren">(</span><em>bwa_path</em>, <em>samtools_path</em>, <em>sambamba_path</em>, <em>reference</em>, <em>output_prefix</em>, <em>fastqs</em>, <em>threads</em>, <em>read_group_line</em>, <em>bwa_params</em>, <em>cram=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/assemble.html#bwa_mem_mapping_sambamba"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#assemble.bwa_mem_mapping_sambamba" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps reads to a reference genome using bwa mem.  If output is in bam format,
will sort using sambamba, else will sort with samtools</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>bwa_path</strong> : str</p>
<blockquote>
<div><p>The path to bwa</p>
</div></blockquote>
<p><strong>samtools_path</strong> : str</p>
<blockquote>
<div><p>The path to samtools</p>
</div></blockquote>
<p><strong>sambamba_path</strong> : str</p>
<blockquote>
<div><p>The path to sambamba</p>
</div></blockquote>
<p><strong>reference</strong> : str</p>
<blockquote>
<div><p>The path to the reference genome (in fasta format)</p>
</div></blockquote>
<p><strong>output_prefix</strong> : str</p>
<blockquote>
<div><p>The prefix (including path) to the desired output files</p>
</div></blockquote>
<p><strong>fastqs</strong> : list</p>
<blockquote>
<div><p>Fastqs, e.g. [&#8216;sample_1.fastq&#8217;, &#8216;sample_2.fastq&#8217;]</p>
</div></blockquote>
<p><strong>threads</strong> : int</p>
<blockquote>
<div><p>The number of threads/cpus to use</p>
</div></blockquote>
<p><strong>read_group_line</strong> : str</p>
<blockquote>
<div><p>Read group info for bwa to add</p>
</div></blockquote>
<p><strong>bwa_params</strong> : list</p>
<blockquote>
<div><p>Bwa parameters to be joined into a string and inserted into the command</p>
</div></blockquote>
<p><strong>cram</strong> : bool</p>
<blockquote>
<div><p>If True, will output a sorted cram, else a sorted bam. Default is False.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">str</p>
<blockquote class="last">
<div><p>Path to output bam file (indexed)</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="utilities">
<h2>Utilities<a class="headerlink" href="#utilities" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="utils.check_bam_fasta_compatibility">
<code class="descclassname">utils.</code><code class="descname">check_bam_fasta_compatibility</code><span class="sig-paren">(</span><em>bam_object</em>, <em>fasta_object</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#check_bam_fasta_compatibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#utils.check_bam_fasta_compatibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks to see if bam and fasta sequence names and lengths are
equivalent (i.e., if it is likely that the bam file was generated
using the fasta in question).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>bam_object</strong> : BamFile() object</p>
<p><strong>fasta_object: RefFasta() object</strong></p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">bool</p>
<blockquote class="last">
<div><p>True if sequence names and lengths match. False otherwise.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="utils.chromosome_wide_plot">
<code class="descclassname">utils.</code><code class="descname">chromosome_wide_plot</code><span class="sig-paren">(</span><em>chrom</em>, <em>positions</em>, <em>y_value</em>, <em>measure_name</em>, <em>sampleID</em>, <em>output_prefix</em>, <em>MarkerSize</em>, <em>MarkerAlpha</em>, <em>Xlim</em>, <em>Ylim</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#chromosome_wide_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#utils.chromosome_wide_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots values across a chromosome, where the x axis is the position along the
chromosome and the Y axis is the value of the measure of interest.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>positions</strong> : numpy array</p>
<blockquote>
<div><p>Genomic coordinates</p>
</div></blockquote>
<p><strong>y_value</strong> : numpy array</p>
<blockquote>
<div><p>The values of the measure of interest</p>
</div></blockquote>
<p><strong>measure_name</strong> : str</p>
<blockquote>
<div><p>The name of the measure of interest (y axis title)</p>
</div></blockquote>
<p><strong>chromosome</strong> : str</p>
<blockquote>
<div><p>The name of the chromosome being plotted</p>
</div></blockquote>
<p><strong>sampleID</strong> : str</p>
<blockquote>
<div><p>The name of the sample</p>
</div></blockquote>
<p><strong>MarkerSize</strong> : float</p>
<blockquote>
<div><p>Size in points^2</p>
</div></blockquote>
<p><strong>MarkerAlpha</strong> : float</p>
<blockquote>
<div><p>Transparency (0 to 1)</p>
</div></blockquote>
<p><strong>Xlim</strong> : float</p>
<blockquote>
<div><p>Maximum X value</p>
</div></blockquote>
<p><strong>Ylim</strong> : float</p>
<blockquote>
<div><p>Maximum Y value</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">int</p>
<blockquote class="last">
<div><p>0</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="utils.chromosome_bed">
<code class="descclassname">utils.</code><code class="descname">chromosome_bed</code><span class="sig-paren">(</span><em>bamfile_obj</em>, <em>output_file</em>, <em>chromosome_list</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#chromosome_bed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#utils.chromosome_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes list of chromosomes and outputs a bed file with the length of each
chromosome on each line (e.g., chr1    0   247249719).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>bamfile_obj</strong> : BamFile() object</p>
<p><strong>output_file</strong> : str</p>
<blockquote>
<div><p>Name of (including full path to) desired output file</p>
</div></blockquote>
<p><strong>chromosome_list</strong> : list</p>
<blockquote>
<div><p>Chromosome/scaffolds to include</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">str</p>
<blockquote>
<div><p>output_file</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>RuntimeError</strong></p>
<blockquote class="last">
<div><p>If chromocomse name is not in bam header.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="utils.make_region_lists">
<code class="descclassname">utils.</code><code class="descname">make_region_lists</code><span class="sig-paren">(</span><em>depthAndMapqDf</em>, <em>mapqCutoff</em>, <em>depth_thresh</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#make_region_lists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#utils.make_region_lists" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters a pandas dataframe for mapq and depth</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>depthAndMapqDf</strong> : pandas dataframe</p>
<blockquote>
<div><p>Must have &#8216;depth&#8217; and &#8216;mapq&#8217; columns</p>
</div></blockquote>
<p><strong>mapqCutoff</strong> : int</p>
<blockquote>
<div><p>The minimum mapq for a window to be considered high quality</p>
</div></blockquote>
<p><strong>depth_thresh</strong> : float</p>
<blockquote>
<div><p>Factor to use in filtering regions based on depth. Li (2014) recommends:
mean_depth +- (depth_thresh * (depth_mean ** 0.5)), where depth_thresh
is 3 or 4.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p>(passing dataframe, failing dataframe)</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="utils.merge_bed_files">
<code class="descclassname">utils.</code><code class="descname">merge_bed_files</code><span class="sig-paren">(</span><em>output_file</em>, <em>*bed_files</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#merge_bed_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#utils.merge_bed_files" title="Permalink to this definition">¶</a></dt>
<dd><p>This function simply takes an output_file (full path to desired output file)
and an arbitrary number of external bed files (including full path),
and merges the bed files into the output_file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>output_file</strong> : str</p>
<blockquote>
<div><p>Full path to and name of desired output bed file</p>
</div></blockquote>
<p><strong>*bed_files</strong></p>
<blockquote>
<div><p>Variable length argument list of external bed files (include full path)</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">str</p>
<blockquote class="last">
<div><p>path to output_file</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="utils.output_bed">
<code class="descclassname">utils.</code><code class="descname">output_bed</code><span class="sig-paren">(</span><em>outBed</em>, <em>*regionDfs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#output_bed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#utils.output_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Concatenate and merges dataframes into an output bed file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>outBed</strong> : str</p>
<blockquote>
<div><p>The full path to and name of the output bed file</p>
</div></blockquote>
<p><strong>*regionDfs</strong></p>
<blockquote>
<div><p>Variable length list of dataframes to be included</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">int</p>
<blockquote class="last">
<div><p>0</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="utils.plot_depth_mapq">
<code class="descclassname">utils.</code><code class="descname">plot_depth_mapq</code><span class="sig-paren">(</span><em>data_dict</em>, <em>output_prefix</em>, <em>sampleID</em>, <em>chrom_length</em>, <em>MarkerSize</em>, <em>MarkerAlpha</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#plot_depth_mapq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#utils.plot_depth_mapq" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates histograms and genome-wide plots of various metrics.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>data_dict</strong> : dict</p>
<blockquote>
<div><p>Key must be &#8216;windows&#8217; value is a pandas data frame</p>
</div></blockquote>
<p><strong>output_prefix</strong> : str</p>
<blockquote>
<div><p>Path and prefix of output files to create</p>
</div></blockquote>
<p><strong>sampleID</strong> : str</p>
<blockquote>
<div><p>Sample ID</p>
</div></blockquote>
<p><strong>chrom_length: int</strong></p>
<blockquote>
<div><p>Length of chromosome</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">int</p>
<blockquote class="last">
<div><p>0</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">API</a><ul>
<li><a class="reference internal" href="#reference-fasta-files">Reference/Fasta files</a></li>
<li><a class="reference internal" href="#bam-files">Bam files</a></li>
<li><a class="reference internal" href="#variant-calling-vcf-processing">Variant Calling/ VCF Processing</a></li>
<li><a class="reference internal" href="#ploidy-estimation">Ploidy Estimation</a></li>
<li><a class="reference internal" href="#mapping-assembly">Mapping/assembly</a></li>
<li><a class="reference internal" href="#utilities">Utilities</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="faqs.html" title="previous chapter">Frequently Asked Questions</a></li>
      <li>Next: <a href="release.html" title="next chapter">Release History</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/api.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Tim Webster, Tanya Phung, Madeline Couse, Bruno Grande, Eric Karlins, Phillip Richmond, Whitney Whitford, Melissa Wilson Sayres.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
      |
      <a href="_sources/api.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>