<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>utils &#8212; XYalign 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for utils</h1><div class="highlight"><pre>
<span></span><span class="c1"># Part of XYalign</span>
<span class="c1"># Collection of miscellaneous functions</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">division</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">print_function</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="c1"># Matplotlib needs to be called in this way to set the display variable</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s1">&#39;Agg&#39;</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="k">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="c1"># Create logger for utils submodule</span>
<span class="n">utils_logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;xyalign.utils&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="chromosome_bed"><a class="viewcode-back" href="../api.html#utils.chromosome_bed">[docs]</a><span class="k">def</span> <span class="nf">chromosome_bed</span><span class="p">(</span><span class="n">bamfile_obj</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="n">chromosome_list</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Takes list of chromosomes and outputs a bed file with the length of each</span>
<span class="sd">	chromosome on each line (e.g., chr1    0   247249719).</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	bamfile_obj : BamFile() object</span>
<span class="sd">	output_file : str</span>
<span class="sd">		Name of (including full path to) desired output file</span>
<span class="sd">	chromosome_list : list</span>
<span class="sd">		Chromosome/scaffolds to include</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	str</span>
<span class="sd">		output_file</span>

<span class="sd">	Raises</span>
<span class="sd">	------</span>

<span class="sd">	RuntimeError</span>
<span class="sd">		If chromocomse name is not in bam header.</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">c_bed_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
	<span class="n">utils_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating bed file with chromosome lengths for </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chromosome_list</span><span class="p">)))</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
		<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chromosome_list</span><span class="p">:</span>
			<span class="k">try</span><span class="p">:</span>
				<span class="n">l</span> <span class="o">=</span> <span class="n">bamfile_obj</span><span class="o">.</span><span class="n">get_chrom_length</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
				<span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="se">\t</span><span class="si">{}</span><span class="se">\t</span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">l</span><span class="p">))</span>
			<span class="k">except</span><span class="p">:</span>
				<span class="n">utils_logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
					<span class="s2">&quot;Error finding chromosome length in bam file </span><span class="si">{}</span><span class="s2"> &quot;</span>
					<span class="s2">&quot;(for bed file)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">bamfile_obj</span><span class="o">.</span><span class="n">filepath</span><span class="p">))</span>
				<span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
					<span class="s2">&quot;Error finding chromosome length in bam file </span><span class="si">{}</span><span class="s2">.  Check &quot;</span>
					<span class="s2">&quot;chromosome names and bam header.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
						<span class="n">bamfile_obj</span><span class="o">.</span><span class="n">filepath</span><span class="p">))</span>
	<span class="n">utils_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Bed file (</span><span class="si">{}</span><span class="s2">) created. Elapsed time: </span><span class="si">{}</span><span class="s2"> seconds&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">output_file</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">c_bed_start</span><span class="p">))</span>
	<span class="k">return</span> <span class="n">output_file</span></div>


<div class="viewcode-block" id="check_bam_fasta_compatibility"><a class="viewcode-back" href="../api.html#utils.check_bam_fasta_compatibility">[docs]</a><span class="k">def</span> <span class="nf">check_bam_fasta_compatibility</span><span class="p">(</span><span class="n">bam_object</span><span class="p">,</span> <span class="n">fasta_object</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Checks to see if bam and fasta sequence names and lengths are</span>
<span class="sd">	equivalent (i.e., if it is likely that the bam file was generated</span>
<span class="sd">	using the fasta in question).</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	bam_object : BamFile() object</span>
<span class="sd">	fasta_object: RefFasta() object</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	bool</span>
<span class="sd">		True if sequence names and lengths match. False otherwise.</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">utils_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
		<span class="s2">&quot;Checking compatibility of </span><span class="si">{}</span><span class="s2"> and </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
			<span class="n">bam_object</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">fasta_object</span><span class="o">.</span><span class="n">filepath</span><span class="p">))</span>

	<span class="n">b_names</span> <span class="o">=</span> <span class="n">bam_object</span><span class="o">.</span><span class="n">chromosome_names</span><span class="p">()</span>
	<span class="n">f_names</span> <span class="o">=</span> <span class="n">fasta_object</span><span class="o">.</span><span class="n">chromosome_names</span><span class="p">()</span>
	<span class="n">b_lengths</span> <span class="o">=</span> <span class="n">bam_object</span><span class="o">.</span><span class="n">chromosome_lengths</span><span class="p">()</span>
	<span class="n">f_lengths</span> <span class="o">=</span> <span class="n">fasta_object</span><span class="o">.</span><span class="n">chromosome_lengths</span><span class="p">()</span>

	<span class="k">if</span> <span class="n">b_lengths</span> <span class="o">==</span> <span class="n">f_lengths</span><span class="p">:</span>
		<span class="k">if</span> <span class="n">b_names</span> <span class="o">==</span> <span class="n">f_names</span><span class="p">:</span>
			<span class="n">utils_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Bam and Fasta are compatible&quot;</span><span class="p">)</span>
			<span class="k">return</span> <span class="kc">True</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">utils_logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
				<span class="s2">&quot;Bam and Fasta are incompatible.  Sequence lengths are identical &quot;</span>
				<span class="s2">&quot;but names are not.  Check chromosome ids to ensure compatibility. &quot;</span>
				<span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> ids are: </span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span>
				<span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> ids are: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
					<span class="n">bam_object</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">b_names</span><span class="p">,</span>
					<span class="n">fasta_object</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">f_names</span><span class="p">))</span>
			<span class="k">return</span> <span class="kc">False</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">utils_logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
			<span class="s2">&quot;Bam and Fasta are incompatible. Check sequence names and lengths &quot;</span>
			<span class="s2">&quot;to ensure correct fasta was used as input. Chromosome ids and &quot;</span>
			<span class="s2">&quot;lengths for </span><span class="si">{}</span><span class="s2"> are:</span><span class="se">\n</span><span class="s2">&quot;</span>
			<span class="s2">&quot;</span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span>
			<span class="s2">&quot;Chromosome ids and lengths for </span><span class="si">{}</span><span class="s2"> are:</span><span class="se">\n</span><span class="s2">&quot;</span>
			<span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
				<span class="n">bam_object</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="nb">zip</span><span class="p">(</span><span class="n">b_names</span><span class="p">,</span> <span class="n">b_lengths</span><span class="p">),</span>
				<span class="n">fasta_object</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="nb">zip</span><span class="p">(</span><span class="n">f_names</span><span class="p">,</span> <span class="n">f_lengths</span><span class="p">)))</span>
		<span class="k">return</span> <span class="kc">False</span></div>


<div class="viewcode-block" id="merge_bed_files"><a class="viewcode-back" href="../api.html#utils.merge_bed_files">[docs]</a><span class="k">def</span> <span class="nf">merge_bed_files</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="o">*</span><span class="n">bed_files</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	This function simply takes an output_file (full path to desired output file)</span>
<span class="sd">	and an arbitrary number of external bed files (including full path),</span>
<span class="sd">	and merges the bed files into the output_file</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	output_file : str</span>
<span class="sd">		Full path to and name of desired output bed file</span>
<span class="sd">	*bed_files</span>
<span class="sd">		Variable length argument list of external bed files (include full path)</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	str</span>
<span class="sd">		path to output_file</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">merged_bed_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
	<span class="n">utils_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging bed files: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bed_files</span><span class="p">)))</span>
	<span class="n">a</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">bed_files</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
	<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">bed_files</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
		<span class="n">b</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">cat</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
	<span class="n">c</span> <span class="o">=</span> <span class="n">b</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span><span class="o">.</span><span class="n">merge</span><span class="p">()</span>
	<span class="n">c</span><span class="o">.</span><span class="n">saveas</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
	<span class="n">utils_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
		<span class="s2">&quot;Merging bed files complete. Elapsed time: </span><span class="si">{}</span><span class="s2"> seconds&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
			<span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_bed_start</span><span class="p">))</span>
	<span class="k">return</span> <span class="n">output_file</span></div>


<div class="viewcode-block" id="make_region_lists"><a class="viewcode-back" href="../api.html#utils.make_region_lists">[docs]</a><span class="k">def</span> <span class="nf">make_region_lists</span><span class="p">(</span><span class="n">depthAndMapqDf</span><span class="p">,</span> <span class="n">mapqCutoff</span><span class="p">,</span> <span class="n">depth_thresh</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Filters a pandas dataframe for mapq and depth</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	depthAndMapqDf : pandas dataframe</span>
<span class="sd">		Must have &#39;depth&#39; and &#39;mapq&#39; columns</span>
<span class="sd">	mapqCutoff : int</span>
<span class="sd">		The minimum mapq for a window to be considered high quality</span>
<span class="sd">	depth_thresh : float</span>
<span class="sd">		Factor to use in filtering regions based on depth. Li (2014) recommends:</span>
<span class="sd">		mean_depth +- (depth_thresh * (depth_mean ** 0.5)), where depth_thresh</span>
<span class="sd">		is 3 or 4.</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	tuple</span>
<span class="sd">		(passing dataframe, failing dataframe)</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">make_region_lists_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
	<span class="n">depth_mean</span> <span class="o">=</span> <span class="n">depthAndMapqDf</span><span class="p">[</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
	<span class="n">depth_sd</span> <span class="o">=</span> <span class="n">depthAndMapqDf</span><span class="p">[</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

	<span class="n">depthMin</span> <span class="o">=</span> <span class="n">depth_mean</span> <span class="o">-</span> <span class="p">(</span><span class="n">depth_thresh</span> <span class="o">*</span> <span class="p">(</span><span class="n">depth_mean</span> <span class="o">**</span> <span class="mf">0.5</span><span class="p">))</span>
	<span class="n">depthMax</span> <span class="o">=</span> <span class="n">depth_mean</span> <span class="o">+</span> <span class="p">(</span><span class="n">depth_thresh</span> <span class="o">*</span> <span class="p">(</span><span class="n">depth_mean</span> <span class="o">**</span> <span class="mf">0.5</span><span class="p">))</span>

	<span class="n">utils_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
		<span class="s2">&quot;Filtering dataframe for mapq (MAPQ &gt;= mapqCutoff) &quot;</span>
		<span class="s2">&quot;and depth (between depthMin and depthMax)&quot;</span><span class="p">)</span>

	<span class="n">good</span> <span class="o">=</span> <span class="p">(</span>
		<span class="p">(</span><span class="n">depthAndMapqDf</span><span class="o">.</span><span class="n">mapq</span> <span class="o">&gt;=</span> <span class="n">mapqCutoff</span><span class="p">)</span> <span class="o">&amp;</span>
		<span class="p">(</span><span class="n">depthAndMapqDf</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">depthMin</span><span class="p">)</span> <span class="o">&amp;</span>
		<span class="p">(</span><span class="n">depthAndMapqDf</span><span class="o">.</span><span class="n">depth</span> <span class="o">&lt;</span> <span class="n">depthMax</span><span class="p">))</span>
	<span class="n">dfGood</span> <span class="o">=</span> <span class="n">depthAndMapqDf</span><span class="p">[</span><span class="n">good</span><span class="p">]</span>
	<span class="n">dfBad</span> <span class="o">=</span> <span class="n">depthAndMapqDf</span><span class="p">[</span><span class="o">~</span><span class="n">good</span><span class="p">]</span>

	<span class="n">utils_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering complete. Elapsed time: </span><span class="si">{}</span><span class="s2"> seconds&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">make_region_lists_start</span><span class="p">))</span>
	<span class="k">return</span> <span class="p">(</span><span class="n">dfGood</span><span class="p">,</span> <span class="n">dfBad</span><span class="p">)</span></div>


<div class="viewcode-block" id="output_bed"><a class="viewcode-back" href="../api.html#utils.output_bed">[docs]</a><span class="k">def</span> <span class="nf">output_bed</span><span class="p">(</span><span class="n">outBed</span><span class="p">,</span> <span class="o">*</span><span class="n">regionDfs</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Concatenate and merges dataframes into an output bed file</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	outBed : str</span>
<span class="sd">		The full path to and name of the output bed file</span>
<span class="sd">	*regionDfs</span>
<span class="sd">		Variable length list of dataframes to be included</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	int</span>
<span class="sd">		0</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">dfComb</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">regionDfs</span><span class="p">)</span>
	<span class="n">regionList</span> <span class="o">=</span> <span class="n">dfComb</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="s2">&quot;chrom&quot;</span><span class="p">:</span><span class="s2">&quot;stop&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
	<span class="n">merge</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">regionList</span><span class="p">)</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span><span class="o">.</span><span class="n">merge</span><span class="p">()</span>
	<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outBed</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output</span><span class="p">:</span>
		<span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">merge</span><span class="p">))</span>
	<span class="k">return</span> <span class="mi">0</span></div>


<div class="viewcode-block" id="chromosome_wide_plot"><a class="viewcode-back" href="../api.html#utils.chromosome_wide_plot">[docs]</a><span class="k">def</span> <span class="nf">chromosome_wide_plot</span><span class="p">(</span>
	<span class="n">chrom</span><span class="p">,</span> <span class="n">positions</span><span class="p">,</span> <span class="n">y_value</span><span class="p">,</span> <span class="n">measure_name</span><span class="p">,</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">,</span>
	<span class="n">MarkerSize</span><span class="p">,</span> <span class="n">MarkerAlpha</span><span class="p">,</span> <span class="n">Xlim</span><span class="p">,</span> <span class="n">Ylim</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Plots values across a chromosome, where the x axis is the position along the</span>
<span class="sd">	chromosome and the Y axis is the value of the measure of interest.</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	positions : numpy array</span>
<span class="sd">		Genomic coordinates</span>
<span class="sd">	y_value : numpy	array</span>
<span class="sd">		The values of the measure of interest</span>
<span class="sd">	measure_name : str</span>
<span class="sd">		The name of the measure of interest (y axis title)</span>
<span class="sd">	chromosome : str</span>
<span class="sd">		The name of the chromosome being plotted</span>
<span class="sd">	sampleID : str</span>
<span class="sd">		The name of the sample</span>
<span class="sd">	MarkerSize : float</span>
<span class="sd">		Size in points^2</span>
<span class="sd">	MarkerAlpha : float</span>
<span class="sd">		Transparency (0 to 1)</span>
<span class="sd">	Xlim : float</span>
<span class="sd">		Maximum X value</span>
<span class="sd">	Ylim : float</span>
<span class="sd">		Maximum Y value</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	int</span>
<span class="sd">		0</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="k">if</span> <span class="s2">&quot;x&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
		<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;green&quot;</span>
	<span class="k">elif</span> <span class="s2">&quot;y&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
		<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;blue&quot;</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;red&quot;</span>
	<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
	<span class="n">axes</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
		<span class="n">positions</span><span class="p">,</span> <span class="n">y_value</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">Color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">MarkerAlpha</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">MarkerSize</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">Xlim</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">Ylim</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> - </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">sampleID</span><span class="p">,</span> <span class="n">chrom</span><span class="p">))</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Chromosomal Position&quot;</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">measure_name</span><span class="p">)</span>
	<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_GenomicScatter.svg&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">output_prefix</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">measure_name</span><span class="p">))</span>
	<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_GenomicScatter.png&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">output_prefix</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">measure_name</span><span class="p">))</span>
	<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
	<span class="k">return</span> <span class="mi">0</span></div>


<div class="viewcode-block" id="plot_depth_mapq"><a class="viewcode-back" href="../api.html#utils.plot_depth_mapq">[docs]</a><span class="k">def</span> <span class="nf">plot_depth_mapq</span><span class="p">(</span>
	<span class="n">data_dict</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">,</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">chrom_length</span><span class="p">,</span> <span class="n">MarkerSize</span><span class="p">,</span> <span class="n">MarkerAlpha</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Creates histograms and genome-wide plots of various metrics.</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	data_dict : dict</span>
<span class="sd">		Key must be &#39;windows&#39; value is a pandas data frame</span>
<span class="sd">	output_prefix : str</span>
<span class="sd">		Path and prefix of output files to create</span>
<span class="sd">	sampleID : str</span>
<span class="sd">		Sample ID</span>
<span class="sd">	chrom_length: int</span>
<span class="sd">		Length of chromosome</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	int</span>
<span class="sd">		0</span>

<span class="sd">	&quot;&quot;&quot;</span>

	<span class="n">window_df</span> <span class="o">=</span> <span class="kc">None</span> <span class="k">if</span> <span class="s2">&quot;windows&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">data_dict</span> <span class="k">else</span> <span class="n">data_dict</span><span class="p">[</span>
		<span class="s2">&quot;windows&quot;</span><span class="p">]</span>
	<span class="n">depth_hist</span> <span class="o">=</span> <span class="kc">None</span> <span class="k">if</span> <span class="s2">&quot;depth_freq&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">data_dict</span> <span class="k">else</span> <span class="n">data_dict</span><span class="p">[</span>
		<span class="s2">&quot;depth_freq&quot;</span><span class="p">]</span>
	<span class="n">readbal_hist</span> <span class="o">=</span> <span class="kc">None</span> <span class="k">if</span> <span class="s2">&quot;readbal_freq&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">data_dict</span> <span class="k">else</span> <span class="n">data_dict</span><span class="p">[</span>
		<span class="s2">&quot;readbal_freq&quot;</span><span class="p">]</span>
	<span class="n">mapq_hist</span> <span class="o">=</span> <span class="kc">None</span> <span class="k">if</span> <span class="s2">&quot;mapq_freq&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">data_dict</span> <span class="k">else</span> <span class="n">data_dict</span><span class="p">[</span>
		<span class="s2">&quot;mapq_freq&quot;</span><span class="p">]</span>

	<span class="n">chromosome</span> <span class="o">=</span> <span class="n">window_df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

	<span class="c1"># Create genome-wide plots based on window means</span>
	<span class="k">if</span> <span class="n">window_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
		<span class="c1"># depth plot</span>
		<span class="c1"># depth_genome_plot_path = os.path.join(</span>
		<span class="c1"># 	output_dir, &quot;depth_windows.&quot; + chromosome + &quot;.png&quot;)</span>
		<span class="c1"># depth_genome_plot = sns.lmplot(</span>
		<span class="c1"># 	&#39;start&#39;, &#39;depth&#39;, data=window_df, fit_reg=False,</span>
		<span class="c1"># 	scatter_kws={&#39;alpha&#39;: 0.3})</span>
		<span class="c1"># depth_genome_plot.savefig(depth_genome_plot_path)</span>
		<span class="n">chromosome_wide_plot</span><span class="p">(</span>
			<span class="n">chromosome</span><span class="p">,</span> <span class="n">window_df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">window_df</span><span class="p">[</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
			<span class="s2">&quot;Depth&quot;</span><span class="p">,</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">,</span>
			<span class="n">MarkerSize</span><span class="p">,</span> <span class="n">MarkerAlpha</span><span class="p">,</span>
			<span class="n">chrom_length</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>

		<span class="c1"># mapping quality plot</span>
		<span class="c1"># mapq_genome_plot_path = os.path.join(</span>
		<span class="c1"># 	output_dir, &quot;mapq_windows.&quot; + chrom + &quot;.png&quot;)</span>
		<span class="c1"># mapq_genome_plot = sns.lmplot(</span>
		<span class="c1"># 	&#39;start&#39;, &#39;mapq&#39;, data=window_df, fit_reg=False)</span>
		<span class="c1"># mapq_genome_plot.savefig(mapq_genome_plot_path)</span>
		<span class="n">chromosome_wide_plot</span><span class="p">(</span>
			<span class="n">chromosome</span><span class="p">,</span> <span class="n">window_df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">window_df</span><span class="p">[</span><span class="s2">&quot;mapq&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
			<span class="s2">&quot;Mapq&quot;</span><span class="p">,</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">,</span>
			<span class="n">MarkerSize</span><span class="p">,</span> <span class="n">MarkerAlpha</span><span class="p">,</span> <span class="n">chrom_length</span><span class="p">,</span> <span class="mi">80</span><span class="p">)</span>

	<span class="c1"># Create histograms</span>
	<span class="c1"># TODO: update filenames dynamically like window_df above</span>
	<span class="c1"># TODO: update Count column name</span>
	<span class="k">if</span> <span class="n">depth_hist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
		<span class="n">depth_bar_plot</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span>
			<span class="n">x</span><span class="o">=</span><span class="s1">&#39;depth&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Count&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">depth_hist</span><span class="p">)</span>
		<span class="n">depth_bar_plot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;depth_hist.png&quot;</span><span class="p">)</span>
	<span class="k">if</span> <span class="n">readbal_hist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
		<span class="n">balance_bar_plot</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span>
			<span class="n">x</span><span class="o">=</span><span class="s1">&#39;ReadBalance&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Count&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">readbal_hist</span><span class="p">)</span>
		<span class="n">balance_bar_plot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;readbalance_hist.png&quot;</span><span class="p">)</span>
	<span class="k">if</span> <span class="n">mapq_hist</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
		<span class="n">mapq_bar_plot</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span>
			<span class="n">x</span><span class="o">=</span><span class="s1">&#39;Mapq&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Count&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">mapq_hist</span><span class="p">)</span>
		<span class="n">mapq_bar_plot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;mapq_hist.png&quot;</span><span class="p">)</span>
	<span class="k">return</span> <span class="mi">0</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Tim Webster, Tanya Phung, Madeline Couse, Bruno Grande, Eric Karlins, Phillip Richmond, Whitney Whitford, Melissa Wilson Sayres.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
    </div>

    

    
  </body>
</html>