

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>variants &mdash; XYalign 0.0.1 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="XYalign 0.0.1 documentation" href="../index.html"/>
        <link rel="up" title="Module code" href="index.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> XYalign
          

          
          </a>

          
            
            
              <div class="version">
                0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faqs.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release.html">Release History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release.html#contributions">Contributions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">XYalign</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
          <li><a href="index.html">Module code</a> &raquo;</li>
      
    <li>variants</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for variants</h1><div class="highlight"><pre>
<span></span><span class="c1"># Part of XYalign</span>
<span class="c1"># Functions for calling and processing variants</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">division</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">bam</span>
<span class="c1"># Matplotlib needs to be called in this way to set the display variable</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s1">&#39;Agg&#39;</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="k">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="c1"># Create logger for variants submodule</span>
<span class="n">variants_logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;xyalign.variants&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="platypus_caller"><a class="viewcode-back" href="../api.html#variants.platypus_caller">[docs]</a><span class="k">def</span> <span class="nf">platypus_caller</span><span class="p">(</span>
	<span class="n">platypus_path</span><span class="p">,</span> <span class="n">log_path</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">chroms</span><span class="p">,</span> <span class="n">cpus</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span>
	<span class="n">regions_file</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Uses platypus to make variant calls on provided bam file</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	platypus_path : str</span>
<span class="sd">		Path to platypus</span>
<span class="sd">	log_path : str</span>
<span class="sd">		Path to and name of desired log file for platypus</span>
<span class="sd">	bam : str</span>
<span class="sd">		Path to input bam (or cram) file</span>
<span class="sd">	ref : str</span>
<span class="sd">		Path to reference sequence</span>
<span class="sd">	chroms : list</span>
<span class="sd">		Chromosomes to call variants on, e.g., [&quot;chrX&quot;, &quot;chrY&quot;, &quot;chr19&quot;]</span>
<span class="sd">	cpus : int</span>
<span class="sd">		Number of threads/cores to use</span>
<span class="sd">	output_file : path</span>
<span class="sd">		Path to and name of the output vcf</span>
<span class="sd">	regions_file : {str, None}</span>
<span class="sd">		If not None, must be path to bed file containing regions to call variants</span>
<span class="sd">		in.  If None, calls in call regions of provided chromosomes. Default =</span>
<span class="sd">		None.</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	int</span>
<span class="sd">		Exit code of the platypus call</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">platy_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
	<span class="k">if</span> <span class="n">regions_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
		<span class="n">regions</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">chroms</span><span class="p">))</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">regions</span> <span class="o">=</span> <span class="n">regions_file</span>
	<span class="n">command_line</span> <span class="o">=</span> <span class="p">[</span>
		<span class="n">platypus_path</span><span class="p">,</span> <span class="s2">&quot;callVariants&quot;</span><span class="p">,</span> <span class="s2">&quot;--bamFiles&quot;</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="s2">&quot;-o&quot;</span><span class="p">,</span>
		<span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;--refFile&quot;</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="s2">&quot;--nCPU&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">cpus</span><span class="p">),</span> <span class="s2">&quot;--regions&quot;</span><span class="p">,</span> <span class="n">regions</span><span class="p">,</span>
		<span class="s2">&quot;--assemble&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;--logFileName&quot;</span><span class="p">,</span> <span class="n">log_path</span><span class="p">]</span>
	<span class="n">variants_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calling variants with command line: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">command_line</span><span class="p">)))</span>
	<span class="n">return_code</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">command_line</span><span class="p">)</span>
	<span class="n">variants_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
		<span class="s2">&quot;Variant calling complete. Elapsed time: </span><span class="si">{}</span><span class="s2"> seconds&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
			<span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">platy_start</span><span class="p">))</span>
	<span class="k">return</span> <span class="n">return_code</span></div>


<div class="viewcode-block" id="parse_platypus_VCF"><a class="viewcode-back" href="../api.html#variants.parse_platypus_VCF">[docs]</a><span class="k">def</span> <span class="nf">parse_platypus_VCF</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">qual_cutoff</span><span class="p">,</span> <span class="n">chrom</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Parse vcf generated by Platypus to grab read balance</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	filename : str</span>
<span class="sd">		Full path to the input vcf</span>
<span class="sd">	qual_cutoff : str</span>
<span class="sd">		Minimum (PHRED) quality at which sites should be included</span>
<span class="sd">	chrom : str</span>
<span class="sd">		Name of the chromosome to include</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>
<span class="sd">	tuple</span>
<span class="sd">		three corresponding arrays of the same length:</span>
<span class="sd">			(position across the chromosome, site quality, read balance)</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">parse_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
	<span class="n">variants_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parsing </span><span class="si">{}</span><span class="s2"> for read balance.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
	<span class="n">infile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">),</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
	<span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">quality</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="n">readBalance</span> <span class="o">=</span> <span class="p">[]</span>
	<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
		<span class="n">cols</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">if</span> <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">chrom</span><span class="p">:</span>
			<span class="k">continue</span>
		<span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
		<span class="n">qual</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">5</span><span class="p">])</span>
		<span class="k">if</span> <span class="n">qual</span> <span class="o">&lt;</span> <span class="n">qual_cutoff</span><span class="p">:</span>
			<span class="k">continue</span>
		<span class="n">TR</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">17</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
		<span class="n">TC</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">14</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
		<span class="k">if</span> <span class="s1">&#39;,&#39;</span> <span class="ow">in</span> <span class="n">TR</span> <span class="ow">or</span> <span class="s1">&#39;,&#39;</span> <span class="ow">in</span> <span class="n">TC</span><span class="p">:</span>
			<span class="k">continue</span>
		<span class="k">if</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">TR</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">TC</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
			<span class="k">continue</span>
		<span class="n">ReadRatio</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">TR</span><span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">TC</span><span class="p">)</span>

		<span class="c1"># Add to arrays</span>
		<span class="n">readBalance</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ReadRatio</span><span class="p">)</span>
		<span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
		<span class="n">quality</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">qual</span><span class="p">)</span>
	<span class="n">variants_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parsing complete. Elapsed time: </span><span class="si">{}</span><span class="s2"> seconds&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">parse_start</span><span class="p">))</span>
	<span class="k">return</span> <span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">quality</span><span class="p">,</span> <span class="n">readBalance</span><span class="p">)</span></div>


<div class="viewcode-block" id="plot_variants_per_chrom"><a class="viewcode-back" href="../api.html#variants.plot_variants_per_chrom">[docs]</a><span class="k">def</span> <span class="nf">plot_variants_per_chrom</span><span class="p">(</span>
	<span class="n">chrom_list</span><span class="p">,</span> <span class="n">vcf_file</span><span class="p">,</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">,</span> <span class="n">qual_cutoff</span><span class="p">,</span>
	<span class="n">MarkerSize</span><span class="p">,</span> <span class="n">MarkerAlpha</span><span class="p">,</span> <span class="n">bamfile_obj</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Parses a vcf file and plots read balance in separate plots</span>
<span class="sd">	for each chromosome in the input list</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	chrom_list : list</span>
<span class="sd">		Chromosomes to include</span>
<span class="sd">	vcf_file : str</span>
<span class="sd">		File (including path) of platypus vcf to analyze</span>
<span class="sd">	sampleID : str</span>
<span class="sd">		Sample ID (for plot titles)</span>
<span class="sd">	output_prefix : str</span>
<span class="sd">		Full path to and prefix of desired output plots</span>
<span class="sd">	qual_cutoff : int</span>
<span class="sd">		Minimum (Phred) quality to consider a site in the vcf</span>
<span class="sd">	MarkerSize : float</span>
<span class="sd">		Size of markers (matplotlib sizes) to use in the figure</span>
<span class="sd">	MarkerAlpha : float</span>
<span class="sd">		Transparency (matplotlib values, 0 to 1) of markers</span>
<span class="sd">	bamfile_obj : BamFile() object</span>
<span class="sd">		Used to get chromosome lengths only</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>
<span class="sd">	int</span>
<span class="sd">		0</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">plot_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
	<span class="n">variants_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Plotting read balance from </span><span class="si">{}</span><span class="s2"> for chroms: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">vcf_file</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chrom_list</span><span class="p">)))</span>
	<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chrom_list</span><span class="p">:</span>
		<span class="n">parse_results</span> <span class="o">=</span> <span class="n">parse_platypus_VCF</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">,</span> <span class="n">qual_cutoff</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
		<span class="n">plot_read_balance</span><span class="p">(</span>
			<span class="n">i</span><span class="p">,</span> <span class="n">parse_results</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">parse_results</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
			<span class="n">sampleID</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">,</span> <span class="n">MarkerSize</span><span class="p">,</span> <span class="n">MarkerAlpha</span><span class="p">,</span> <span class="n">bamfile_obj</span><span class="p">)</span>
		<span class="n">hist_read_balance</span><span class="p">(</span>
			<span class="n">i</span><span class="p">,</span> <span class="n">parse_results</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">)</span>
	<span class="n">variants_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
		<span class="s2">&quot;Read balance plotting complete. Elapsed time: </span><span class="si">{}</span><span class="s2"> seconds&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
			<span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">plot_start</span><span class="p">))</span>
	<span class="k">return</span> <span class="mi">0</span></div>


<div class="viewcode-block" id="plot_read_balance"><a class="viewcode-back" href="../api.html#variants.plot_read_balance">[docs]</a><span class="k">def</span> <span class="nf">plot_read_balance</span><span class="p">(</span>
	<span class="n">chrom</span><span class="p">,</span> <span class="n">positions</span><span class="p">,</span> <span class="n">readBalance</span><span class="p">,</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">,</span> <span class="n">MarkerSize</span><span class="p">,</span>
	<span class="n">MarkerAlpha</span><span class="p">,</span> <span class="n">bamfile_obj</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Plots read balance at each SNP along a chromosome</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	chrom : str</span>
<span class="sd">		Name of the chromosome</span>
<span class="sd">	positions : numpy array</span>
<span class="sd">		Positions along the chromosome (same length as readBalance)</span>
<span class="sd">	readBalance : numpy array</span>
<span class="sd">		Read balance corresponding with the positions in the positions array</span>
<span class="sd">	sampleID : str</span>
<span class="sd">		Sample name or id to include in the plot title</span>
<span class="sd">	output_prefix : str</span>
<span class="sd">		Desired prefix (including full path) of the output files</span>
<span class="sd">	MarkerSize : float</span>
<span class="sd">		Size of markers (matplotlib sizes) to use in the figure</span>
<span class="sd">	MarkerAlpha : float</span>
<span class="sd">		Transparency (matplotlib values) of markers for the figure</span>
<span class="sd">	bamfile_obj : BamFile() object</span>
<span class="sd">		Used to get chromosome lengths only</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>
<span class="sd">	int</span>
<span class="sd">		0</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">chrom_len</span> <span class="o">=</span> <span class="n">bamfile_obj</span><span class="o">.</span><span class="n">get_chrom_length</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
	<span class="k">if</span> <span class="s2">&quot;x&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
		<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;green&quot;</span>
	<span class="k">elif</span> <span class="s2">&quot;y&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
		<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;blue&quot;</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;red&quot;</span>
	<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
	<span class="n">axes</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
		<span class="n">positions</span><span class="p">,</span> <span class="n">readBalance</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">Color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">MarkerAlpha</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">MarkerSize</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">chrom_len</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">sampleID</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Chromosomal Coordinate&quot;</span><span class="p">)</span>
	<span class="n">axes</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Read Balance&quot;</span><span class="p">)</span>
	<span class="c1"># print(len(positions))</span>
	<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_ReadBalance_GenomicScatter.svg&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">output_prefix</span><span class="p">,</span> <span class="n">chrom</span><span class="p">))</span>
	<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_ReadBalance_GenomicScatter.png&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">output_prefix</span><span class="p">,</span> <span class="n">chrom</span><span class="p">))</span>
	<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
	<span class="n">variants_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Genomic read balance plot of </span><span class="si">{}</span><span class="s2"> complete.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
		<span class="n">chrom</span><span class="p">))</span>
	<span class="k">return</span> <span class="mi">0</span></div>


<div class="viewcode-block" id="hist_read_balance"><a class="viewcode-back" href="../api.html#variants.hist_read_balance">[docs]</a><span class="k">def</span> <span class="nf">hist_read_balance</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">readBalance</span><span class="p">,</span> <span class="n">sampleID</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Plots a histogram of read balance</span>

<span class="sd">	Parameters</span>
<span class="sd">	----------</span>

<span class="sd">	chrom : str</span>
<span class="sd">		Name of the chromosome</span>
<span class="sd">	readBalance : nump array</span>
<span class="sd">		Read balance values</span>
<span class="sd">	sampleID : str</span>
<span class="sd">		Sample name or id to include in the plot title</span>
<span class="sd">	output_prefix : str</span>
<span class="sd">		Desired prefix (including full path) of the output files</span>

<span class="sd">	Returns</span>
<span class="sd">	-------</span>

<span class="sd">	int</span>
<span class="sd">		0</span>

<span class="sd">	&quot;&quot;&quot;</span>
	<span class="k">try</span><span class="p">:</span>
		<span class="n">read_balance</span> <span class="o">=</span> <span class="n">readBalance</span><span class="p">[</span><span class="mf">0.05</span> <span class="o">&lt;</span> <span class="n">readBalance</span> <span class="o">&lt;</span> <span class="mf">0.95</span><span class="p">]</span>
		<span class="k">if</span> <span class="s2">&quot;x&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
			<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;green&quot;</span>
		<span class="k">elif</span> <span class="s2">&quot;y&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="o">.</span><span class="n">lower</span><span class="p">():</span>
			<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;blue&quot;</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">Color</span> <span class="o">=</span> <span class="s2">&quot;red&quot;</span>
		<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
		<span class="n">axes</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>
		<span class="n">axes</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">sampleID</span><span class="p">)</span>
		<span class="n">axes</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Read Balance&quot;</span><span class="p">)</span>
		<span class="n">axes</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
		<span class="n">axes</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">readBalance</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">Color</span><span class="p">)</span>
		<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_ReadBalance_Hist.svg&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">,</span> <span class="n">chrom</span><span class="p">))</span>
		<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_ReadBalance_Hist.png&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">,</span> <span class="n">chrom</span><span class="p">))</span>
		<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
		<span class="n">variants_logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Genomic read balance histogram of </span><span class="si">{}</span><span class="s2"> complete.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
			<span class="n">chrom</span><span class="p">))</span>
	<span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
		<span class="n">variants_logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
			<span class="s2">&quot;No sites on </span><span class="si">{}</span><span class="s2"> to plot histogram. Skipping.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>
	<span class="k">return</span> <span class="mi">0</span></div>
</pre></div>

           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Tim Webster, Tanya Phung, Madeline Couse, Bruno Grande, Eric Karlins, Phillip Richmond, Whitney Whitford, Melissa Wilson Sayres.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'0.0.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>